### PR TITLE
Fix sprite size error and rendering error

### PR DESCRIPTION
**Pull request by cptbtptpbcptdtptp**
`renderer.position` should not be coupled with `sprite.texture`

Repro code
### The default size of sprites is not updated
```
import {
  AssetType,
  Camera,
  Rect,
  Sprite,
  Texture2D,
  Vector3,
  WebGLEngine,
} from "@galacean/engine";

// Create engine
WebGLEngine.create({ canvas: "canvas" }).then((engine) => {
  engine.canvas.resizeByClientSize();

  const scene = engine.sceneManager.activeScene;
  const rootEntity = scene.createRootEntity();

  const cameraEntity = rootEntity.createChild("Camera");
  cameraEntity.transform.position = new Vector3(0, 0, 50);
  cameraEntity.addComponent(Camera);

  engine.resourceManager
    .load<Texture2D>({
      url: "https://gw.alipayobjects.com/mdn/rms_7c464e/afts/img/A*KjnzTpE8LdAAAAAAAAAAAAAAARQnAQ",
      type: AssetType.Texture2D,
    })
    .then((texture) => {
      const sprite = new Sprite(engine, texture);
      console.log("pre sprite.width", sprite.width);
      sprite.region = new Rect(0, 0, 0.5, 1);
      console.log("aft sprite.width", sprite.width);
    });

  engine.run();
});
```

### An exception is displayed when setting the texture asynchronously
```
import {
  AssetType,
  Camera,
  Sprite,
  SpriteRenderer,
  Texture2D,
  Vector3,
  WebGLEngine,
} from "@galacean/engine";

// Create engine
WebGLEngine.create({ canvas: "canvas" }).then((engine) => {
  engine.canvas.resizeByClientSize();

  const scene = engine.sceneManager.activeScene;
  const rootEntity = scene.createRootEntity();

  const cameraEntity = rootEntity.createChild("Camera");
  cameraEntity.transform.position = new Vector3(0, 0, 50);
  cameraEntity.addComponent(Camera);

  const sprite = new Sprite(engine);
  const spriteRenderer = rootEntity.addComponent(SpriteRenderer);
  spriteRenderer.sprite = sprite;
  spriteRenderer.width = 10;
  spriteRenderer.height = 10;

  engine.resourceManager
    .load<Texture2D>({
      url: "https://gw.alipayobjects.com/mdn/rms_7c464e/afts/img/A*KjnzTpE8LdAAAAAAAAAAAAAAARQnAQ",
      type: AssetType.Texture2D,
    })
    .then((texture) => {
      sprite.texture = texture;
    });

  engine.run();
});
```

### CODE REVIEW

**Comment by cptbtptpbcptdtptp** [comment]
Fix the problem that the default width and height are not corrected after the sprite modifies the region and other attributes.
```
const sprite = new Sprite(engine);
// sprite size, undefined, undefined
console.log("sprite size", sprite.width, sprite.height);
engine.resourceManager
  .load<Texture2D>({
    url: "https://gw.alipayobjects.com/mdn/rms_7c464e/afts/img/A*d3N9RYpcKncAAAAAAAAAAAAAARQnAQ",
    type: AssetType.Texture2D,
  })
  .then((texture) => {
    sprite.texture = texture;
    // sprite size, 3.2, 3.2
    console.log("sprite size", sprite.width, sprite.height);
    sprite.region = new Rect(0, 0, 1, 0.5);
    // Did not meet expectations !!!
    // sprite size, 3.2, 3.2
    console.log("sprite size", sprite.width, sprite.height);
  });
```

File: packages/core/src/2d/sprite/Sprite.ts [hunk 1]
```diff
@@ -44,21 +46,27 @@ export class Sprite extends RefObject {
...
     }
   }
 
   /**
    * The width of the sprite (in world coordinates).
    */
   get width(): number {
-    this._width === undefined && this._calDefaultSize();
-    return this._width;
+    if (this._customWidth !== undefined) {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/Sprite.ts:59]: According to the current trend, we can no longer use unspoken rules:
1. First of all, you need to write clearly what the "trigger conditions" of the two modes are
2. What is the difference between the two

File: packages/core/src/2d/sprite/Sprite.ts [hunk 2]
```diff
@@ -67,13 +75,17 @@ export class Sprite extends RefObject {
    * The height of the sprite (in world coordinates).
    */
   get height(): number {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/Sprite.ts:77]: The same with `width`

File: packages/core/src/2d/sprite/Sprite.ts [hunk 3]
```diff
@@ -44,21 +46,27 @@ export class Sprite extends RefObject {
...
+      if (this._customWidth === undefined || this._customHeight === undefined) {
+        this._dispatchSpriteChange(SpriteModifyFlags.size);
+      }
     }
   }
 
   /**
    * The width of the sprite (in world coordinates).
    */
   get width(): number {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/Sprite.ts:58]: According to the current trend, we can no longer use unspoken rules:
1. First of all, you need to write clearly what the "trigger conditions" of the two modes are
2. What is the difference between the two

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 4]
```diff
@@ -38,10 +38,16 @@ export class SpriteRenderer extends Renderer implements ICustomClone {
   @ignoreClone
   private _sprite: Sprite = null;
 
+  @assignmentClone
+  private _customWidth: number = undefined;
+  @assignmentClone
+  private _customHeight: number = undefined;
   @ignoreClone
-  private _width: number = undefined;
+  private _width: number = 0;
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:46]: same advice as `Sprite`

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 5]
```diff
@@ -117,15 +124,17 @@ export class SpriteRenderer extends Renderer implements ICustomClone {
    * Render width.
    */
   get width(): number {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:126]: same advice as Sprite

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 6]
```diff
@@ -38,10 +38,16 @@ export class SpriteRenderer extends Renderer implements ICustomClone {
...
+  @assignmentClone
+  private _customHeight: number = undefined;
   @ignoreClone
-  private _width: number = undefined;
+  private _width: number = 0;
   @ignoreClone
-  private _height: number = undefined;
+  private _height: number = 0;
+  @ignoreClone
+  private _defaultSizeDirty: boolean = true;
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:50]: Why not use `_dirtyUpdateFlag`?

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 7]
```diff
@@ -216,6 +227,7 @@ export class SpriteRenderer extends Renderer implements ICustomClone {
    */
   _cloneTo(target: SpriteRenderer): void {
     target.sprite = this._sprite;
+    target.drawMode = this._drawMode;
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:230]: Unexpected delivery?😄

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 8]
```diff
@@ -274,6 +286,16 @@ export class SpriteRenderer extends Renderer implements ICustomClone {
     }
   }
 
+  private _calDefaultSize() {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:289]: add `:void`

Comment (GuoLei1990) [reply packages/core/src/2d/sprite/SpriteRenderer.ts:289]:
```ts
  private _calDefaultSize(): void {
    const sprite = this._sprite;
    if (sprite) {
      this._width = sprite.width;
      this._height = sprite.height;
    } else {
      this._width = this._height = 0;
    }
    this._defaultSizeDirty = false;
  }
```

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 9]
```diff
@@ -304,9 +326,14 @@ export class SpriteRenderer extends Renderer implements ICustomClone {
         this.shaderData.setTexture(SpriteRenderer._textureProperty, this.sprite.texture);
         break;
       case SpriteModifyFlags.size:
+        this._defaultSizeDirty = true;
         // When the width and height of `SpriteRenderer` are `undefined`,
         // the `size` of `Sprite` will affect the position of `SpriteRenderer`.
-        if (this._drawMode === SpriteDrawMode.Sliced || this._width === undefined || this._height === undefined) {
+        if (
+          this._drawMode === SpriteDrawMode.Sliced ||
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:333]: May tiled mode should add too?

**Comment by GuoLei1990** [comment]
- `SpriteMask` is same as suggested `SpriteRenderer`, No longer！

- You need to add a complete unit test to make the coverage of Sprite, SpriteRenderer, and SpriteMask 100%

**Review by GuoLei1990 — CHANGES_REQUESTED** [review]
little change, good!

File: packages/core/src/2d/sprite/Sprite.ts (hunk 2 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/2d/sprite/Sprite.ts:79]: Done.

File: packages/core/src/2d/sprite/Sprite.ts [hunk 10]
```diff
@@ -12,8 +12,10 @@ export class Sprite extends RefObject {
   /** The name of sprite. */
   name: string;
 
-  private _width: number = undefined;
-  private _height: number = undefined;
+  private _width: number = 0;
+  private _height: number = 0;
+  private _customWidth: number = undefined;
+  private _customHeight: number = undefined;
```
[the remark this answers is not in the document]
Comment (cptbtptpbcptdtptp) [reply packages/core/src/2d/sprite/Sprite.ts:18]: Done

File: packages/core/src/2d/sprite/Sprite.ts (hunk 3 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/2d/sprite/Sprite.ts:59]: Done

File: packages/core/src/2d/sprite/Sprite.ts [hunk 11]
```diff
@@ -340,5 +366,6 @@ export class Sprite extends RefObject {
 enum SpriteUpdateFlags {
   positions = 0x1,
   uvs = 0x2,
-  all = 0x3
+  defaultSize = 0x4,
```
[the remark this answers is not in the document]
Comment (cptbtptpbcptdtptp) [reply packages/core/src/2d/sprite/Sprite.ts]: Done

File: packages/core/src/2d/sprite/SpriteRenderer.ts (hunk 4 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/2d/sprite/SpriteRenderer.ts]: Done.

File: packages/core/src/2d/sprite/SpriteRenderer.ts (hunk 5 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/2d/sprite/SpriteRenderer.ts:123]: Done.

File: packages/core/src/2d/sprite/SpriteRenderer.ts (hunk 6 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/2d/sprite/SpriteRenderer.ts]: Done.

File: packages/core/src/2d/sprite/SpriteRenderer.ts (hunk 7 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/2d/sprite/SpriteRenderer.ts:228]: Yes.

File: packages/core/src/2d/sprite/SpriteRenderer.ts (hunk 8 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/2d/sprite/SpriteRenderer.ts]: Done.

**Review by GuoLei1990 — APPROVED** [review]

**Pull request merged**